scipy/stats/_distn_infrastructure.py

```diff
@@ -1283,21 +1291,45 @@ def moment(self, n, *args, **kwds):
                 mdict = {'moments': {1: 'm', 2: 'v', 3: 'vs', 4: 'vk'}[n]}
             else:
                 mdict = {}
-            mu, mu2, g1, g2 = self._stats(*args, **mdict)
-        val = _moment_from_stats(n, mu, mu2, g1, g2, self._munp, args)
+            mu, mu2, g1, g2 = self._stats(*shapes, **mdict)
+        val = np.empty(loc.shape)  # val needs to be indexed by loc
+        val[...] = _moment_from_stats(n, mu, mu2, g1, g2, self._munp, shapes)
 
         # Convert to transformed  X = L + S*Y
         # E[X^n] = E[(L+S*Y)^n] = L^n sum(comb(n, k)*(S/L)^k E[Y^k], k=0...n)
-        if loc == 0:
-            return scale**n * val
-        else:
-            result = 0
-            fac = float(scale) / float(loc)
+        result = zeros(i0.shape)
+        place(result, ~i0, self.badvalue)
+
+        if i1.any():
+            res1 = scale[loc == 0]**n * val[loc == 0]
+            place(result, i1, res1)
+
+        if i2.any():
+            mom = [mu, mu2, g1, g2]
+            arrs = [i for i in mom if i is not None]
```

Comment:
Was the problem that we assumed if `mu` was `None` then the rest would be `None`?

Could you explain what is going on here?

Reply:
> Was the problem that we assumed if mu was None then the rest would be None?

Yes, I think so.

> Could you explain what is going on here?

I am creating a list of moments that are not `None` and calling `argsreduce` to remove the invalid entries. But as we don't know which of them would be `None` and which wouldn't, I have also created a list of indices recording the index of the moments that are not `None`. Once `argsreduce` returns a list of reduced moments, I just fill in the `mom` list with the moments that have been reduced (using the index list).

I didn't think much about a better approach to get this done quickly. If this is acceptable for now, it should work just fine and we can get back to it later and do something that is clearer. Does that sound good?

Reply:
> Would you reach out to some other core developers who might be interested?

Sure.

> I'll take a look at some of those PRs you submitted for adding alternative to mstats hypothesis tests.

Thanks!